tests/tests_generator.cpp

```diff
@@ -0,0 +1,379 @@
+////////////////////////////////////////////////////////////////
+// Tests adapted from reference implementation of std::generator proposal P2502R2
+// https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2022/p2502r2.pdf
+
+#define DOCTEST_CONFIG_IMPLEMENT_WITH_MAIN
+#include <doctest/doctest.h>
+
+#include <dice/template-library/generator.hpp>
+
+#include <array>
+#include <cstdint>
+#include <cstdio>
+#include <string>
+#include <tuple>
+#include <vector>
+#include <type_traits>
+
+TEST_SUITE("generator") {
+	using namespace DICE_TEMPLATELIBRARY_GENERATOR_NAMESPACE;
+
+	///////////////////////
+    // Simple non-nested serial generator
+
+    generator<uint64_t> fib(int max) {
+        auto a = 0, b = 1;
+        for (auto n = 0; n < max; n++) {
+            co_yield std::exchange(a, std::exchange(b, a + b));
+        }
+    }
+
+    generator<int> other_generator(int i, int j) {
+        while (i != j) {
+            co_yield i++;
+        }
+    }
+
+    /////////////////////
+    // Demonstrate ability to yield nested sequences
+    //
+    // Need to use std::ranges::elements_of() to trigger yielding elements of
+    // nested sequence.
+    //
+    // Supports yielding same generator type (with efficient resumption for
+    // recursive cases)
+    //
+    // Also supports yielding any other range whose elements are convertible to
+    // the current generator's elements.
+
+    generator<uint64_t, uint64_t> nested_sequences_example() {
+        std::printf("yielding elements_of std::array\n");
+#if defined(__GNUC__) && !defined(__clang__)
+        co_yield ranges::elements_of(std::array<const uint64_t, 5>{2, 4, 6, 8, 10}, {});
+#else
+        co_yield ranges::elements_of{std::array<const uint64_t, 5>{2, 4, 6, 8, 10}};
+#endif
+        std::printf("yielding elements_of nested std::generator\n");
+        co_yield ranges::elements_of{fib(10)};
+
+        std::printf("yielding elements_of other kind of generator\n");
+        co_yield ranges::elements_of{other_generator(5, 8)};
+    }
+
+    //////////////////////////////////////
+    // Following examples show difference between:
+    //
+    //                                    If I co_yield a...
+    //                              X / X&&  | X&         | const X&
+    //                           ------------+------------+-----------
+    // - generator<X, X>               (same as generator<X, X&&>)
+    // - generator<X, const X&>   ref        | ref        | ref
+    // - generator<X, X&&>        ref        | ill-formed | ill-formed
+    // - generator<X, X&>         ill-formed | ref        | ill-formed
+
+    struct X {
+        int id;
```

Comment:
I don't think thats possible just because NRVO is still optional and may or may not come into play. So I don't think we can correctly determine what is supposed to happen across all compilers